xapian-applications/omega/assistant.cc

```diff
@@ -0,0 +1,127 @@
+/** @file assistant.cc
+ * @brief Worker module for putting text extraction into a separate process.
+ */
+/* Copyright (C) 2011 Olly Betts
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+#include <csignal>
+
+#include "worker_comms.h"
+#include "handler_pdf.h"
+#include "safeunistd.h"
+
+using namespace std;
+
+const int FD = 3;
+const int time_limit = 300;
+
+#if defined SIGALRM
+// IF I have SIGALRM (Best option)
+
+static void
+timeout_handler(int n) {
+    exit(n);
+}
+
+static void
+set_timeout() {
+    // Causes the system to generate a SIGALRM signal for the process after
+    // time_limit of seconds
+    alarm(time_limit);
+    // Set a signal handler for SIGALRM
+    signal(SIGALRM, timeout_handler);
+}
+
+static void
+stop_timeout() {
+    // Set SIGALRM to be ignore
+    signal(SIGALRM, SIG_IGN);
+    // A pending alarm request, if any, is cancelled by calling alarm(0)
+    alarm(0);
+}
+
+// #elif defined PTHREAD?
+/* If I can use threads (second option)
```

Comment:
It is possible to use conditional compilation. But I am not sure about which variables I should use, or how to create new ones.

Reply:
I guess (as above) you probably can skip this threaded code at least for now, but to answer the question:

With autotools, there are build-time checks in `configure.ac` to see what features are available, and this arranges for `HAVE_xxx` macros to be defined or not (or in a few cases to be defined to `0` or `1`) to indicate the results of these checks.

If you need threads, we require at least C++11 now which includes thread support so that's an option too.

Reply:
Okay, I will take a look at it and try to add a variable for the timer.